File: robots/ros/src/tutorial/scripts/publisher.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>

#include <sstream>

using namespace ros;

int main (int argc, char **argv) {
    init(argc, argv, "publisher_cpp");
    NodeHandle nh;

    Publisher publisher = nh.advertise<std_msgs::String>("chatter_cpp", 10);

    Rate rate(10);
    int count = 0;

    while (ok()) {
        std_msgs::String msg;

        std::stringstream ss;
        ss << "hello from C++ " << count;
        msg.data = ss.str();

        ROS_INFO("%s", msg.data.c_str());
        publisher.publish(msg);

        spinOnce();
        rate.sleep();
        count++;
    }

    return 0;
}

File: robots/ros/src/tutorial/scripts/publisher.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import String

def main():
    rospy.init_node('publisher')
    publisher = rospy.Publisher('chatter_py', String, queue_size=10)
    rate = rospy.Rate(2)  # 2 Hz

    while not rospy.is_shutdown():
        msg = String()
        msg.data = "Hello, ROS!"
        publisher.publish(msg)
        rate.sleep()

    rospy.loginfo("Publisher node is shutting down.")

if __name__ == '__main__':
    main()